### PR TITLE
CDAP-1961: Initial cube dataset implementation

### PR DESCRIPTION
**Pull request by abaranau**
Initial CubeDataset implementation based on existing Cube APIs (no changes there).

### CODE REVIEW

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT1390-2 GREEN

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CubeModule.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+public class CubeModule implements DatasetModule {
+  public static final String SHORT_NAME = "cube";
+  public static final String FULL_NAME = Cube.class.getName();
+
+  @Override
+  public void register(DatasetDefinitionRegistry registry) {
+    DatasetDefinition<Table, ? extends DatasetAdmin> tableDef = registry.get("table");
+    registry.add(new CubeDatasetDefinition(FULL_NAME, tableDef));
+    registry.add(new CubeDatasetDefinition(SHORT_NAME, tableDef));
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CubeModule.java]: why do we need to add it under two names ?

Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CubeModule.java]: we do that for all datasets. Let's keep consistency

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+          throw new IllegalArgumentException("Invalid property: " + prop.getKey());
+        }
+      }
+    }
+
+    List<Aggregation> aggregations = Lists.newArrayList();
+    for (Map.Entry<String, List<String>> aggTagsEntry : aggTags.entrySet()) {
+      Set<String> requiredTags = aggRequiredTags.get(aggTagsEntry.getKey());
+      requiredTags = requiredTags == null ? Collections.<String>emptySet() : requiredTags;
+      aggregations.add(new DefaultAggregation(aggTagsEntry.getValue(), requiredTags));
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: what if aggTagsEntry.getValue() is null ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    String resProp = propsMap.get(PROPERTY_RESOLUTIONS);
+    int[] resolutions;
+    if (resProp == null) {
+      resolutions = DEFAULT_RESOLUTIONS;
+    } else {
+      String[] seconds = resProp.split(",");
+      resolutions = new int[seconds.length];
+      for (int i = 0; i < seconds.length; i++) {
+        resolutions[i] = Integer.valueOf(seconds[i]);
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: what if resolution is > INTEGER_MAX ? should we fail only for that resolution or all of them ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+ * Aggregation is defined with list of tags to aggregate by and a list of required tags
+ * (dataset.cube.aggregation.[agg_name].tags and dataset.cube.aggregation.[agg_name].requiredTags properties
+ * respectively). The {@link co.cask.cdap.api.dataset.lib.cube.CubeFact} measurement is aggregated within an aggregation
+ * if it contains all required tags which non-null value.
+ */
+public class CubeDatasetDefinition extends AbstractDatasetDefinition<CubeDataset, DatasetAdmin> {
+  public static final String PROPERTY_RESOLUTIONS = "dataset.cube.resolutions";
+  public static final String PROPERTY_AGGREGATION_PREFIX = "dataset.cube.aggregation.";
+  public static final String PROPERTY_TAGS = "tags";
+  public static final String PROPERTY_REQUIRED_TAGS = "requiredTags";
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: do we need to have the above properties configurable from cdap-site ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/MetricsTableOnTable.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  @Override
+  public void put(NavigableMap<byte[], NavigableMap<byte[], Long>> updates) throws Exception {
+    for (Map.Entry<byte[], NavigableMap<byte[], Long>> rowUpdate : updates.entrySet()) {
+      Put put = new Put(rowUpdate.getKey());
+      for (Map.Entry<byte[], Long> columnUpdate : rowUpdate.getValue().entrySet()) {
+        put.add(columnUpdate.getKey(), columnUpdate.getValue());
+      }
+      table.put(put);
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/MetricsTableOnTable.java]: question : should we batch these puts ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CubeModule.java (hunk 1 above)
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CubeModule.java]: ok

**Comment by gsps1** [comment]
just couple of comments and questions.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java (hunk 2 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: it cannot be null, right?
``` java
 aggTags.put(nameAndProp[0], Arrays.asList(tags));
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java (hunk 3 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: I guess, you'll get NumberFormatException. Let me add handling for that

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java (hunk 4 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: no - those are property names for specific dataset

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java (hunk 2 above)
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDatasetDefinition.java]: yeah, i was thinking if we have required but not aggregate tags, but we are iterating over aggregate tags, you are right.

**Comment by gsps1** [comment]
LGTM

**Pull request merged**